### PR TITLE
Created Fido Client for Gong synoptic maps

### PR DESCRIPTION
**Pull request by abhijeetmanhas**
### Description

Fixes https://github.com/dstansby/pfsspy/issues/165

This will enable users to search and download the meged data product for GONG using Fido, giving the timerange. The url pattern is like this,

`https://gong2.nso.edu/oQR/zqs/%Y%m/mrzqs%y%m%d/mrzqs%y%m%dt%H%Mc\d4_\d3.fits.gz`
Here `\d4` and `\d3` are Carringtion rotation number and Initital carr-long for synoptic map respectively.

Used scraper partially, I can't write client only using scraper since it scraper needs everything to be defined in the pattern, which won't be the case for these two attributes. It doesn't supports custom regex.

### CODE REVIEW

**Comment by abhijeetmanhas** [comment]
Main things to bee reviewed:

- The name of Client
- The `_makemap` routine, what more can be added ?
- Should add its own fetch function, which will extract the files too?
- Add a function to convert the fits data to pfsspy input compatible ?
- Should allow users to specify the `CAR_ROT` and `LONG0` ?
- `_can_handle_query` , what more `phsyobs`, `level`, etc should it support ?

Sunpy scraper should be able to handle custom regex, something by which we can solve issue https://github.com/sunpy/sunpy/issues/3337 by making scraper selecting only those goes_satellites for which data is present. Can be handled in different PR.

**Comment by dstansby** [comment]
> Main things to bee reviewed:
> 
>     * The name of Client
Seems fine to me
> 
>     * The `_makemap` routine, what more can be added ?
> 
I'm not sure what this is, so pass...
>     * Should add its own fetch function, which will extract the files too?
> 
Do you mean un-zip them from `.fits.gz` to `.fits`?
>     * Add a function to convert the fits data to pfsspy input compatible ?
> 
That's out of scope for sunpy, so no
>     * Should allow users to specify the `CAR_ROT` and `LONG0` ?
> 
I think searching by time is fine
>     * `_can_handle_query` , what more `phsyobs`, `level`, etc should it support ?
> 
Again, just searching by time is fine. We can always expand this later.
> 
> Sunpy scraper should be able to handle custom regex, something by which we can solve issue #3337 by making scraper selecting only those goes_satellites for which data is present. but it can be handled in different PR.

**Comment by abhijeetmanhas** [comment]
Yes, I was thinking of unzipping them into fits. Should we do it here? Or just leave gz files downloaded?

**Comment by dstansby** [comment]
Ideally we want to unzip them, if possible.

**Comment by Cadair** [comment]
I am pretty sure the SRS client handles unpacking if you want some prior art.

**Comment by abhijeetmanhas** [comment]
@Cadair thanks for the suggestion.
I discovered your comment after pushing. SRS uses Tarfile library, I am using gzip. 
Which one would be better here?

**Comment by abhijeetmanhas** [comment]
@dstansby the test failure is remote server error, rest all tests have passed. It is good for review now.

File: sunpy/net/dataretriever/sources/gong_synoptic.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        -------
+        boolean
+            answer as to whether client can service the query
+        """
+        chkattr = ['Time', 'Instrument']
+        chklist = [x.__class__.__name__ in chkattr for x in query]
+        for x in query:
+            if x.__class__.__name__ == 'Instrument' and x.value.lower() == 'gong':
+                return all(chklist)
+        return False
```
Comment (dstansby) [inline sunpy/net/dataretriever/sources/gong_synoptic.py:157]: It is possible to download different types of data from GONG, so I think we need to put some more thought into making `_can_handle_query` more specific. I think what we might need is an attribute that specifies we are looking for a synoptic map (as opposed to full disk observation), and an attribute that specifies that we are looking for magnetic field (more specifically, radial magnetic field) data. I think I will open this in a new issue, so people can see/comment on ideas for these attributes (and whether we need to make any new ones).

**Comment by dstansby** [comment]
Thanks for all the work on this! On the technical side it looks 👍. As I've said in a comment, more thought needs to go in on the sunpy/Fido API side to come up with a more specific way to search for this data. I will open a new issue to discuss that.

**Comment by nabobalis** [comment]
Would a specific attr help?

File: sunpy/net/dataretriever/sources/gong_synoptic.py (hunk 1 above)
Comment (abhijeetmanhas) [reply sunpy/net/dataretriever/sources/gong_synoptic.py]: Thanks for the reivew, I have added physobs and extentType now.

**Comment by abhijeetmanhas** [comment]
@nabobalis I have added optional `ExtentType` attrs like satellite no. in goes to differentiate b/w `FULL_DISK` and `SYNOPTIC` data, and tests for the same too. Also I have added optional `physobs` which should by `LOS_MAGNETIC_FIELD` as discussed with @dstansby .

File: sunpy/net/dataretriever/sources/gong_synoptic.py (hunk 1 above)
Comment (abhijeetmanhas) [reply sunpy/net/dataretriever/sources/gong_synoptic.py]: @dstansby this is green now, can be reviewed, I have done all suggested changes till now.

File: sunpy/net/dataretriever/sources/gong_synoptic.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+                    soup = BeautifulSoup(opn, "html.parser")
+                    for link in soup.find_all("a"):
+                        href = link.get("href")
+                        if re.match(file_pattern, href):
+                            # extracting timestamp from the filename
+                            timestring = href[5:16]
+                            tstamp = parse_time('20'+timestring.upper()+'00')
+                            if tstamp >= timerange.start and tstamp <= timerange.end:
+                                fullpath = directory + href
+                                filesurls.append(fullpath)
```
Comment (dpshelio) [inline sunpy/net/dataretriever/sources/gong_synoptic.py:78]: Shouldn't all that be "done" by the scrapper? I believe you can put a wildcard if you don't know how it goes (like the last digits)... Ok, you can't! but with my new pr #4107

File: sunpy/net/dataretriever/sources/gong_synoptic.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+        for url, filename in zip(urls, paths):
+            downloader.enqueue_file(url, filename=filename)
+
+        resobj = downloader.download()
+
+        for loc_fpath, fpath in zip(local_paths, paths):
+            f = gzip.open(fpath, 'r')
+            g = gzip.open(loc_fpath, 'w')
+            g.write(f.read())
```
Comment (dpshelio) [inline sunpy/net/dataretriever/sources/gong_synoptic.py:126]: Why is this needed?

Comment (dpshelio) [reply sunpy/net/dataretriever/sources/gong_synoptic.py:126]: Aren't you removing the `.gz` on line 112? why are you then zipping it back without the `gz` suffix?

Comment (abhijeetmanhas) [reply sunpy/net/dataretriever/sources/gong_synoptic.py:126]: Yes, I removed `.gz` in line 112 to just get what the filename of extracted file should be. I haven't done any extraction there.
`local_fpath` contains path to future extracted file, which is currently empty when I opened it in write mode, and write `f` object to it, which contains the extracted file object in `fits`.

File: sunpy/net/dataretriever/sources/gong_synoptic.py (hunk 2 above)
Comment (abhijeetmanhas) [reply sunpy/net/dataretriever/sources/gong_synoptic.py:78]: This PR will be very helpful. I'll change the code to depend on #4107 after it is merged.

File: sunpy/net/dataretriever/sources/gong_synoptic.py (hunk 3 above)
Comment (dpshelio) [reply sunpy/net/dataretriever/sources/gong_synoptic.py:126]: sorry I've not commented earlier. I don't understand why you want to unzip the file. I believe fits, and map, reads zipped files fine. I'd prefer to keep the files zipped in my computer.

Comment (abhijeetmanhas) [reply sunpy/net/dataretriever/sources/gong_synoptic.py:126]: Cool, keeping it zipped then. Thanks for the review!

File: changelog/4055.feature.rst [hunk 4]
```diff
@@ -0,0 +1,2 @@
+Created `~sunpy.net.dataretriever.sources.GongSynopticClient` for accessing merged data product
```
Comment (nabobalis) [inline changelog/4055.feature.rst:1]: What is a merged data product?

Comment (abhijeetmanhas) [reply changelog/4055.feature.rst]: Changed to 
```Created `~sunpy.net.dataretriever.sources.GongSynopticClient` for accessing Magnetogram Synoptic Map Archives of NSO-GONG.```.

**Comment by abhijeetmanhas** [comment]
`helio` Test failure is not related to the change.

File: sunpy/net/dataretriever/sources/tests/test_gong_synoptic.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    "time, instrument",
+    [(a.Time('2019/10/4', '2019/10/4 2:00'), a.Instrument('gong')),
+     (a.Time('2019/12/31 21:00', '2020/1/1'), a.Instrument('gong'))])
+def test_fido(time, instrument):
+    qr = Fido.search(time, instrument)
+    assert isinstance(qr, UnifiedResponse)
+    response = Fido.fetch(qr)
+    assert len(response) == qr._numfile
+    for res in response:
+        assert res.split('.')[-1] == 'gz'
```
Comment (dpshelio) [inline sunpy/net/dataretriever/sources/tests/test_gong_synoptic.py:99]: I don't really like `assert` on loops, if one fails, you only get a "report" of that one, and not whether others did too.
```suggestion
    assert all(map(lambda x: x.endswith('.gz'), response))
```

File: sunpy/net/dataretriever/sources/gong_synoptic.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+from sunpy.util.scraper import Scraper
+
+__all__ = ['GongSynopticClient']
+
+
+class GongSynopticClient(GenericClient):
+    """
+    Provides access to the Magnetogram products of NSO-GONG synoptic Maps.
+
+    Searches data hosted by the `National Solar Observatory <gong2.nso.edu/oQR/zqs/>`__
```
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/gong_synoptic.py:11]:
```suggestion
    Searches data hosted by the `National Solar Observatory. <gong2.nso.edu/oQR/zqs/>`__
```

File: sunpy/net/dataretriever/sources/gong_synoptic.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+    Examples
+    --------
+
```
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/gong_synoptic.py:15]:
```suggestion
```

File: sunpy/net/dataretriever/sources/gong_synoptic.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    """
+
+    def _get_url_for_timerange(self, timerange, **kwargs):
+        """
+        Return URL(s) for corresponding timerange.
+
+        Parameters
+        ----------
+        timerange : `~sunpy.time.TimeRange`
+            The time range you want the files for.
```
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/gong_synoptic.py:41]:
```suggestion
            The time range required.
```

File: sunpy/net/dataretriever/sources/gong_synoptic.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        """
+        Answers whether client can service the query.
+
+        Parameters
+        ----------
+        query : list of query objects
+
+        Returns
+        -------
+        boolean
```
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/gong_synoptic.py:72]:
```suggestion
        bool : 
```

File: sunpy/net/dataretriever/sources/gong_synoptic.py [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+            answer as to whether client can service the query
```
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/gong_synoptic.py:73]:
```suggestion
            Answer as to whether client can service the query,
```

File: sunpy/net/dataretriever/sources/gong_synoptic.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        """
+        from sunpy.net import attrs as a
```
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/gong_synoptic.py:75]: Does this need to be in the function?

Comment (abhijeetmanhas) [reply sunpy/net/dataretriever/sources/gong_synoptic.py:75]: I thought it prevents circular imports.

Comment (nabobalis) [reply sunpy/net/dataretriever/sources/gong_synoptic.py:75]: Can you check?

Comment (abhijeetmanhas) [reply sunpy/net/dataretriever/sources/gong_synoptic.py:73]: yes, it prevents. https://github.com/sunpy/sunpy/blob/master/sunpy/net/dataretriever/sources/goes.py#L380 similar to what we did for `goes` and `norh`.

File: sunpy/net/dataretriever/sources/gong_synoptic.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+        if not cls.check_attr_types_in_query(query, required, optional):
+            return False
+
+        matches = True
+        for x in query:
+            if isinstance(x, a.Instrument) and x.value.lower() != 'gong':
+                matches = False
+            if isinstance(x, a.Physobs) and x.value.lower() != 'los_magnetic_field':
+                matches = False
+            if isinstance(x, a.gong_synoptic.ExtentType) and x.value.lower() != 'synoptic':
```
Comment (Cadair) [inline sunpy/net/dataretriever/sources/gong_synoptic.py:86]: Why do we have an attr for a thing we only support one type of?

Comment (dstansby) [reply sunpy/net/dataretriever/sources/gong_synoptic.py:86]: I was thinking that we could re-use the VSO `EXTENT_TYPE` keyword here, but it's currently broken: https://github.com/sunpy/sunpy/issues/4048

Comment (Cadair) [reply sunpy/net/dataretriever/sources/gong_synoptic.py:86]: If we don't have any use for this attr I suggest we don't implement it at this time. The extent type for this new client should be set in the `_map` metadata in the client class.

Comment (abhijeetmanhas) [reply sunpy/net/dataretriever/sources/gong_synoptic.py:86]: I added extentType after this https://github.com/sunpy/sunpy/pull/4055#discussion_r414496406 . We can either remove it too.

Comment (Cadair) [reply sunpy/net/dataretriever/sources/gong_synoptic.py:86]: Is gong data with different extent types available through the VSO? but we can't filter on it because of #4048 ?

Comment (dstansby) [reply sunpy/net/dataretriever/sources/gong_synoptic.py:86]: No, I don't think GONG synoptic maps are on the VSO, but I think it would be good if we can re-purpose the EXTENT_TYPE attr in VSO for search (ie. move it from a vso specific attr to a general one like `Time` or `Instrument`) to avoid creating new attrs.

Comment (dstansby) [reply sunpy/net/dataretriever/sources/gong_synoptic.py:86]: Okay, I did some digging/research and the EXTENT_TYPE attr for VSO is properly broken, so I think including a new attr is fine here.

If it was working then these are the valid values: https://sdac.virtualsolar.org/cgi/show_details?keyword=EXTENT_TYPE - @Cadair do you think it's worth putting the new attr in `sunpy.net.attrs` instead of in `sunpy.net.gong_synoptic.attrs`?

File: sunpy/net/dataretriever/sources/gong_synoptic.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+    >>> results  #doctest: +REMOTE_DATA
+    <sunpy.net.fido_factory.UnifiedResponse object at ...>
+    Results from 1 Provider:
+    <BLANKLINE>
+    3 Results from the GongSynopticClient:
+         Start Time           End Time      Source Instrument Wavelength
+    ------------------- ------------------- ------ ---------- ----------
+    2019-12-31 22:00:00 2020-01-01 02:00:00    NSO       GONG        nan
+    2019-12-31 22:00:00 2020-01-01 02:00:00    NSO       GONG        nan
+    2019-12-31 22:00:00 2020-01-01 02:00:00    NSO       GONG        nan
```
Comment (dstansby) [inline sunpy/net/dataretriever/sources/gong_synoptic.py:27]: This result looks odd, do you know why this is picking up three files, but all with the same timestamp?

File: sunpy/net/dataretriever/sources/gong_synoptic.py (hunk 12 above)
Comment (dstansby) [reply sunpy/net/dataretriever/sources/gong_synoptic.py:86]: Going to resolve this conversation, and move discussion on where this new attr should go to an issue.

**Comment by dstansby** [comment]
@nabobalis which PR is this waiting on?

**Comment by nabobalis** [comment]
> @nabobalis which PR is this waiting on?

https://github.com/sunpy/sunpy/pull/4321

**Comment by Cadair** [comment]
#4321 has been merged so this also needs a rewrite.

File: sunpy/net/dataretriever/sources/gong_synoptic.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    def _attrs_module(cls):
+        return 'gong_synoptic', 'sunpy.net.dataretriever.attrs.gong_synoptic'
+
+    @classmethod
+    def register_values(cls):
+        from sunpy.net import attrs
+        adict = {attrs.Instrument: [("GONG", "Global Oscillation Network Group.")],
+                 attrs.Physobs: [("LOS_MAGNETIC_FIELD", "Line of sight magnetic field")],
+                 attrs.Source: [('NSO', 'National Solar Observatory.')],
+                 attrs.gong_synoptic.ExtentType: [("SYNOPTIC", "Coverage of a complete solar rotation synthesized over time")]}
```
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/gong_synoptic.py:45]: I don't have a better suggestion but `attrs.gong_synoptic.ExtentType` seems a bit long.

File: sunpy/net/dataretriever/sources/gong_synoptic.py (hunk 13 above)
Comment (abhijeetmanhas) [reply sunpy/net/dataretriever/sources/gong_synoptic.py]: Fixed now.

File: sunpy/net/dataretriever/sources/gong_synoptic.py (hunk 14 above)
Comment (abhijeetmanhas) [reply sunpy/net/dataretriever/sources/gong_synoptic.py:45]: Can we just remove it?
It will only happen that there won't be an extra column of name `extent type` in results whoes value is "synoptic". 
Nothing else would change.

Comment (dstansby) [reply sunpy/net/dataretriever/sources/gong_synoptic.py:45]: I think it would be reasonable to move `ExtentType` out of `attrs.gong_synoptic` and into `attrs` as a more generic attribute, don't know what others think about that?

Comment (nabobalis) [reply sunpy/net/dataretriever/sources/gong_synoptic.py:45]: IF we need to use it for this client, I would be in favour of that instead of it being for this client only.

Comment (dstansby) [reply sunpy/net/dataretriever/sources/gong_synoptic.py:45]: Given one can also get full disc magnetograms from GONG, I actually think we should rename the client from `gong_synoptic` > `gong`, and keep this keyword as a way to differentiate between a full disk map and a synoptic map. Then we avoid having separate `gong` and a `gong_synoptic` clients.

Comment (abhijeetmanhas) [reply sunpy/net/dataretriever/sources/gong_synoptic.py]: Replaced gongsynoptic with gong and moved extentType to attrs.

File: sunpy/net/attrs.py [hunk 15]
```diff
@@ -20,10 +20,12 @@
 * `a.vso <sunpy.net.vso.attrs>`
 * `a.jsoc <sunpy.net.jsoc.attrs>`
 * `a.goes <sunpy.net.dataretriever.attrs.goes>`
+* `a.gong_synoptic <sunpy.net.dataretriever.attrs.gong_synoptic>`
```
Comment (nabobalis) [inline sunpy/net/attrs.py:23]:
```suggestion
```

**Review by nabobalis — APPROVED** [review]

**Review by kakirastern — APPROVED** [review]

**Review by dstansby — APPROVED** [review]
